File: pos_to_nerf.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <json/json.h>
#include <yaml-cpp/yaml.h>
#include <opencv2/opencv.hpp>

#define CAMCHAIN_FILE "2023-05-03-23-11-22-camchain.yaml"
#define IMAGE_FILE "output/images/test/*.png"
#define POSE_FILE "pos.json"

typedef std::vector<std::vector<double>> Matrix;

Matrix camchain2nerf(const std::string& filepath) {
    YAML::Node data = YAML::LoadFile(filepath);
    Matrix intrin(4, std::vector<double>(4, 0.0));
    Matrix dist;

    intrin[0][0] = data["cam0"]["intrinsics"][0].as<double>();
    intrin[1][1] = data["cam0"]["intrinsics"][1].as<double>();
    intrin[0][2] = data["cam0"]["intrinsics"][2].as<double>();
    intrin[1][2] = data["cam0"]["intrinsics"][3].as<double>();

    const YAML::Node& dist_coeffs = data["cam0"]["distortion_coeffs"];
    for (const auto& coeff : dist_coeffs) {
        dist.push_back({ coeff.as<double>() });
    }

    return std::make_pair(intrin, dist);
}

std::vector<std::string> images2nerf(const std::string& image_file_paths) {
    std::vector<std::string> image_paths;
    cv::glob(image_file_paths, image_paths);
    cv::Mat cv_img = cv::imread(image_paths[0]);
    int h = cv_img.rows;
    int w = cv_img.cols;

    return std::make_pair(image_paths, std::make_pair(h, w));
}

Matrix pos2nerf(const std::string& json_file) {
    std::ifstream file(json_file);
    Json::Value json_pos;
    file >> json_pos;

    Matrix Plist;

    const Json::Value& poses = json_pos["poses"];
    const Json::Value& orientations = json_pos["orientations"];

    for (int i = 0; i < orientations.size(); ++i) {
        Matrix ori = orientations[i].as<Matrix>();
        ori.push_back({ 0, 0, 0 });
        Matrix pos = poses[i].as<Matrix>();
        pos.push_back({ 1 });

        Matrix P(ori.size(), std::vector<double>(ori[0].size() + pos[0].size(), 0.0));
        for (int row = 0; row < ori.size(); ++row) {
            for (int col = 0; col < ori[row].size(); ++col) {
                P[row][col] = ori[row][col];
            }
        }
        for (int row = 0; row < pos.size(); ++row) {
            for (int col = 0; col < pos[row].size(); ++col) {
                P[row][col + ori[row].size()] = pos[row][col];
            }
        }

        Plist.push_back(P);
    }

    return Plist;
}

void integrate() {
    std::map<std::string, double> data_format;
    Matrix intrin, dist;
    std::tie(intrin, dist) = camchain2nerf(CAMCHAIN_FILE);

    double k1 = dist[0][0];
    double k2 = dist[1][0];
    double p1 = dist[2][0];
    double p2 = dist[3][0];
    double fl_x = intrin[0][0];
    double fl_y = intrin[1][0];
    double cx = intrin[2][0];
    double cy = intrin[3][0];

    std::vector<std::string> file_paths;
    int h, w;
    std::tie(file_paths, std::tie(h, w)) = images2nerf(IMAGE_FILE);

    double camera_angle_x = 2 * std::tanh(2 * fl_x / w);
    double camera_angle_y = 2 * std::tanh(2 * fl_y / h);

    data_format["camera_angle_x"] = camera_angle_x;
    data_format["camera_angle_y"] = camera_angle_y;
    data_format["fl_x"] = fl_x;
    data_format["fl_y"] = fl_y;
    data_format["k1"] = k1;
    data_format["k2"] = k2;
    data_format["p1"] = p1;
    data_format["p2"] = p2;
    data_format["cx"] = cx;
    data_format["cy"] = cy;
    data_format["w"] = w;
    data_format["h"] = h;

    Matrix Plist = pos2nerf(POSE_FILE);
    std::vector<std::map<std::string, std::string>> frame_list;
    for (int i = 0; i < file_paths.size(); ++i) {
        std::map<std::string, std::string> frame;
        frame["file_path"] = file_paths[i];
        frame["transform_matrix"] = Json::FastWriter().write(Plist[i]);
        frame_list.push_back(frame);
    }
    data_format["frames"] = frame_list;

    Json::Value json_data;
    for (const auto& entry : data_format) {
        json_data[entry.first] = entry.second;
    }

    std::ofstream output_file("data_format_half.json");
    output_file << json_data;
    output_file.close();

File: pos_to_nerf.py
import json
import yaml
import numpy as np  
import glob
import cv2
from collections import OrderedDict
from pprint import pprint
CAMCHAIN_FILE ="metadata/2023-05-03-23-11-22-camchain.yaml"
IMAGE_FILE = 'output/images/test/*.png'
POSE_FILE = "metadata/pos.json"

def camchain2nerf(filepath):
    with open(filepath, 'r') as f:
        data = {}
        data = yaml.load(f,Loader= yaml.FullLoader)
        intrin = data['cam0']['intrinsics']
        
        dist = np.array((data['cam0']['distortion_coeffs']))

    return intrin,dist

       

def images2nerf(image_file_paths):
    image_paths = sorted(glob.glob(image_file_paths))
    cv_img = cv2.imread(image_paths[0])
    h, w, _ = cv_img.shape

    return image_paths,h,w
    

def pos2nerf(json_file): 
    with open(json_file,'r') as f:
            Plist = []
            json_pos= json.load(f)

            poses = json_pos["poses"]
            orientations = json_pos["orientations"]
            

            for ori, pos in zip(orientations, poses):
                ori = np.vstack((ori,[0,0,0]))
                pos = np.vstack((pos,[1]))
                P = np.hstack((ori,pos))
                Plist.append(P)
    return Plist

               



def intergrate():
    data_format = OrderedDict()
    intrin, dist = camchain2nerf(CAMCHAIN_FILE)
    k1 = dist[0]
    k2 = dist[1]
    p1 = dist[2]
    p2 = dist[3]

    fl_x = intrin[0]
    fl_y = intrin[1]
    cx = intrin[2]
    cy = intrin[3]

    file_paths ,h,w = images2nerf(IMAGE_FILE)
    camera_angle_x = 2 * np.tanh(2*fl_x / w)
    camera_angle_y = 2 * np.tanh(2*fl_y / h)

    data_format["camera_angle_x"] = camera_angle_x
    data_format["camera_angle_y"] = camera_angle_y
    data_format["fl_x"] = fl_x
    data_format["fl_y"] = fl_y
    data_format["k1"] = k1
    data_format["k2"] = k2
    data_format["p1"] = p1
    data_format["p2"] = p2
    data_format["cx"] = cx
    data_format["cy"] = cy
    data_format["w"] = w
    data_format["h"] =h 
    
    # frames 
    Plist = pos2nerf(POSE_FILE)
    frame_list = []
    for file_path,p in zip(file_paths, Plist):
        frame = dict()
        frame["file_path"] =file_path
        #frame["rotation"] = "rotation" 
        frame["transform_matrix"] = p.tolist()
        frame_list.append(frame)
    data_format["frames"] = frame_list
    
    # data_format = dict(data_format)

    return data_format


        
def dict2json(data):
    with open("metadata/data_format_half.json","w") as f:
        json.dump(data,f)

    

    

    
if __name__ =="__main__":

    ret = intergrate()
    dict2json(ret)
    

